### ISSUE TITLE
Do we own the memory pointed to by the returned screenshots? Also: Warning about using the screenshots with OpenCV!

### ISSUE THREAD

**Issue by Arcitec** [SerpentAI/D3DShot#11]
Hey, great library, huge thanks for all of your great work!

So, I'm curious... When we grab images, it's internally done by asking Windows via COM to make a screenshot, and then return the pointer to us...

What does D3DShot do with that pointer? Does it copy the data to our own internal memory within Python, or do we continue to use the Windows pointer to memory owned by Windows?

The reason why I ask is that I've seen some really strange behavior. Perhaps a bug in d3dshot?

```python
import cv2
import d3dshot
import numpy as np

d = d3dshot.create(capture_output="numpy")
img = d.screenshot()
print(img.flags.writeable)
img[:,:,1] = 255 # Write maximum green into every pixel (makes img hue green)
#img = img.copy() # You must uncomment this, otherwise cv2 can't draw on it.
cv2.rectangle(img, (0,0), (100,100), (0,0,255), 6)

cv2.imshow("img", img)
cv2.waitKey(0)
cv2.destroyAllWindows()
```

As the comment/code explains:

- The Numpy array returned by `d.screenshot()` **is** marked as writeable.
- Writing directly into the array with `img[:,:,1] = 255` works perfectly (fills the green channel with 255 in every pixel).
- Passing the img Numpy array to `cv2.rectangle` does NOT draw on the image.
- The only way to draw on the image is to FIRST copy it to a brand new Numpy object and THEN pass the NEW Numpy object to OpenCV.

This is very strange. It seems kinda like a bug. I really can't understand why the Numpy object is marked as writeable and we're able to write to it manually, but OpenCV cannot write to it unless we first copy it. So clearly SOMETHING differs between these objects, but I have no idea what...

---

Actually, I might have just solved it while writing this ticket, thanks to knowing quite a lot about OpenCV internals... Okay, if you look at `print(img.flags)`, the raw screenshot says "CONTIGUOUS = FALSE" which means that the underlying RAM is not laid out in the same way as what the Numpy array is claiming. (You can read more about RAM layout and Contiguous in my article here: https://answers.opencv.org/question/219040/fastest-way-to-convert-bgr-rgb-aka-do-not-use-numpy-magic-tricks/)

So, the following code shows us what is going on:

```python
img = d.screenshot()
print(img.flags) # Contiguous: False, Writeable: True
print(img.strides) # (1920, 1, 2073600)
img = img.copy() # Makes a new, contiguous object
print(img.strides) # (5760, 3, 1)
```

Okay, the next part of understanding this problem is to know about what OpenCV does internally when it receives a Python object. That's described in the link I gave above. Basically, if you give OpenCV a non-contiguous Numpy object, it can't read the RAM as-is. Instead, it creates a new Numpy object with contiguous RAM storage and then reads THAT object's RAM data, and passes it into the internal OpenCV function (ie draw rectangle).

So, OpenCV is taking the screenshot Numpy object (which is incorrectly laid out in RAM), and then it is creating a new (temporary) Numpy object with proper layout, and then it passes that object to the internal C++ based drawing function.

In other words, we're drawing on a temporary object.

Here's the proof:

```python
import cv2
import d3dshot
import numpy as np
import time

d = d3dshot.create(capture_output="numpy")
img = d.screenshot()
#img = img.copy() # Uncomment this to check speed difference!
start = time.perf_counter()
cv2.rectangle(img, (0,0), (100,100), (0,0,255), 6)
#cv2.dnn.blobFromImage(img)
elapsed = (time.perf_counter() - start) * 1000
print("elapsed:", elapsed, "ms")

cv2.imshow("img", img)
cv2.waitKey(0)
cv2.destroyAllWindows()
```

Results: If we're passing the original image given by d3dshot into `cv2.rectangle`, the costly conversion happens in OpenCV's "Python-to-OpenCV" wrapper... and I am getting `148.2 ms` runtime for a call to `cv2.rectangle`. If we instead first do `img = img.copy()` to make a new, contiguous Numpy array, the `cv2.rectangle` completes in just `0.09 ms`.

I also checked with `blobFromImage` and the time was `37 ms` if `img.copy()` was first used, and `46 ms` otherwise (meaning with the bad RAM data layout). So, around +9 ms to call the blob converter when given badly laid out RAM. Note that the "9ms" is no coincidence. That's the average amount of time it takes to run `img = img.copy()`, which confirms that `blobFromImage` has to create a fixed (contiguous) temporary, internal copy of the Numpy array when you call that function.

So yes, it is now confirmed:

**D3DShot Numpy Arrays are HARMFUL if used with OpenCV. _Every_ time you call ANY OpenCV function and give it the original d3dshot Numpy array, you are causing a MASSIVE slowdown since 
 the data has to be internally converted by OpenCV to a _temporary_ Numpy array with corrected memory layout.**

What's the solution then? Well, there are a few solutions. All involve various ways to pre-build a contiguous ndarray so that _OpenCV_ doesn't have to do any conversion when you give it the ndarray later.

- `img = img.copy()`, which takes 9.3 milliseconds on my i7-8750H Laptop at 1920x1080.
- `img = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)` to actually use the OpenCV color converter. This takes 11.3 milliseconds consistently on my computer (of that, 9.3 milliseconds is the time it takes Numpy to copy the non-contiguous data to a contiguous one first, which happens internally in OpenCV when given non-contiguous ndarrays). And this method is obviously only useful if you WANT to change the channel order.
- Perhaps it can be fixed in D3DShot itself, by making it lay out the data contiguously in RAM? I doubt it, but it's worth asking. If we're doing a memory copy of the temporary data that Windows gave us over COM, then perhaps we can do that copy in a way that lays it out contiguously in RAM. That way, the Numpy arrays would be valid for OpenCV API usage, without needing conversion overhead.

**Comment by Arcitec** [comment]
Okay I can answer the first of the two questions: **Do we own the screenshot's memory? Yes.**

`img.flags` says `OWNDATA : False`, which means that Numpy is telling us that the data it points at is not owned by the Numpy object.

But Numpy arrays have a `.base` property pointing to the array that a view-array is based on, if any.

And we can see that `img = d.screenshot(); print(img.base.flags)` says `OWNDATA : True`, which means that the images returned by D3DShot are a non-contiguous view (re-arranged interpretation) of another Numpy array which DOES own its data (but the `img.base.shape` was laid out as `(3, 1080, 1920)` which means color, height, width..., a very unusual order, which is why the secondary "view" had to be created to interpret the data as `(height, width, rgb)` instead).

So yeah, we do own the RAM pointed to by the returned images. That's a relief. Windows won't overwrite it or hate us if we write to it, etc...

---

As for the second issue (the fact that the returned image/ndarray view is non-contiguous, which is a huge performance penalty): That's still unresolved.

**Comment by Arcitec** [comment]
Okay I was looking in the Numpy importer in d3dshot:

https://github.com/SerpentAI/D3DShot/blob/e2801d29dfeb4ccd8252d57749549d0af17c41ca/d3dshot/capture_outputs/numpy_capture_output.py#L15-L29

That's well-written and obviously does the right things:

1. Allocate an empty (non-initialized) ndarray of the exact size in bytes that we'll need, which is a super fast operation.
2. Copy the Windows API's memory as-is to the ndarray in memory, which is a super fast operation.
3. Reshape numpy's interpretation of the array data into `(height, width, channels)`, along with handling rotations of the data whenever the screen is rotated. All of this is pure Numpy view-manipulation (it changes the `strides` and `shape` properties of the Numpy array without editing any data in actual RAM at all). So it's an efficient way to fit square pegs into round holes, by just telling Numpy to reinterpret the actual RAM layout as if it was something else.

So I can see that it would be impossible to fix this problem in D3DShot itself. Because we'd get into the territory of doing a TON of separate `ctypes.memmove` calls to copy small segments of bytes at a time, to somehow put it in the proper order in the final Numpy array, and that'd be super slow.

Therefore, the final answer to the performance issue, is exactly what I said at the end of the initial post above:

1. If you want the data as-is in RGB order, use `img = d.screenshot(); img = img.copy()` to force Numpy to make a new object with the data copied into proper order instead of all the `bullsh*t` memory layout Numpy tricks. This is a pretty slow operation, but it's implemented in C and is therefore way faster than any Python-based solutions would be.
2. If you want the data in BGR order or something else, use `img = d.screenshot(); img = cv2.cvtColor(img, cv2.COLOR_RGB2<Your desired color>)`, which internally causes an `img.copy()` to happen (copying the non-contiguous Numpy data into contiguous RAM), and then converts the color on the copy, and finally returns the fixed copy.

Both of those will result in fast, contiguous RAM image objects. Usable as high-performance objects in further OpenCV (or other libraries) APIs.

Final note: You're always going to pay the cost of converting non-contiguous data to contiguous. There is no such thing as calling an OpenCV (or Matplotlib or similar) API without a conversion (in the case of OpenCV) or slow interpretation of the data (in the case of Matplotlib) happening under the hood. Therefore, the best thing a user can do is to manually and immediately FIX the screenshot data via one of the two methods above.

But the fix is a bit slow, since Numpy is slow at copying-and-fixing non-contiguous RAM, so it's not something that should be a default in the library itself. And whether non-contiguous RAM will cause you huge performance issues or not depends ENTIRELY on what you plan to do with the screenshot data. If you don't plan to manipulate the data via OpenCV or other graphics libraries, and you only want to do numpy math on it, then there's probably no reason to fix the memory layout.

It all depends on the user... In MOST cases, however, users DO want to fix the memory layout. If they don't fix it (via one of the two methods above), it will just slow down all of their various graphics library API calls that they give the bad, non-contiguous data to...

Tough situation, but yeah... I'll resort to always doing `img = d.screenshot(); img = img.copy()` to fix this problem.

**Comment by nbrochu** [comment]
Thanks for the write-up, you found a lot of the stuff I also found out while building D3DShot. I was looking at all cost to avoid that extra copy whenever possible to maximize performance.

I wouldn't say that in the majority of cases users want the ordered memory layout though. Passing D3DShot's NumPy output to scikit-image or Pillow works fine without a copy call. You got me doubting now that maybe the `rot90` arrays could always be forced to copy when moving to an image library though so I'm going to have to test.

Oh and I had encountered a similar problem too with my [PyTorch output](https://github.com/SerpentAI/D3DShot/blob/master/d3dshot/capture_outputs/pytorch_capture_output.py#L16). As you can see, I have `copy()` calls in there because PyTorch doesn't understand Numpy's `rot90` shenanigans. Funnily enough, it has no issue with `reshape`, it just chokes on negative strides it seems.

This doesn't seem like an issue I can truly act on however. Like you saw in your deep dive, it gets pretty wild. I could work on documenting this aspect a little more however.

**Comment by nbrochu** [comment]
Found something super interesting. Turns out it's possible to have the data be contiguous while still being faster than the current method that uses views.

See: https://answers.opencv.org/question/219040/fastest-way-to-convert-bgr-rgb-aka-do-not-use-numpy-magic-tricks/

So for BGRA <=> RGB I ditch `frame[..., [2, 1, 0]]` and go for `cv2.cvtColor(frame, cv2.COLOR_BGRA2RGB)` instead. This shaves off about 7 ms per processed frame AND the data is contiguous!

Downside: I now need opencv-python as a new dependency. I think it's worth it.

**Comment by Arcitec** [comment]
Hi! I've been inactive and didn't see this until now. Thanks for your great replies! I am glad to hear that you've put so much work into optimizing this library! 👍 

> I wouldn't say that in the majority of cases users want the ordered memory layout though. Passing D3DShot's NumPy output to scikit-image or Pillow works fine without a copy call.

It depends on the library.

- OpenCV: Every time you call it with non-contiguous data, it uses the raw Numpy C++ "COPY" function and creates a temporary contiguous copy internally, which is super slow. It's the same as the user manually doing `img.copy()` EVERY time they call any OpenCV function with the screenshot data.
- Matplotlib: It can handle non-contiguous data but it becomes very slow since it has to traverse the RAM memory data in weird ways with weird loops. It's faster to first create clean data via `img.copy()` from the screenshot, before you start giving the data to Matplotlib functions. This avoids triggering Matplotlib's weird, slow, internal "non-contiguous data handling" loop techniques.
- Pillow and scikit-image: I haven't checked what they do internally. Probably same as Matplotlib. Since even if a library supports non-contiguous image data, they still have to traverse such messy data using crazy loops and backwards/forwards strides, which slows down all library calls. If the input image data's RAM layout is in clean, contiguous order, the C++ library code can operate in a straight forward loop from start to end without a ton of nested loops/jumps.

> Found something super interesting.

Hehe thanks, that's my article and a lot of work went into it. I wrote it after going very deeply into optimizing my Python program which uses D3DShot to grab the screen and then analyzes it with neural networks.

> So for BGRA <=> RGB I ditch `frame[..., [2, 1, 0]]` and go for `cv2.cvtColor(frame, cv2.COLOR_BGRA2RGB)` instead. This shaves off about 7 ms per processed frame AND the data is contiguous!
> 
> Downside: I now need opencv-python as a new dependency. I think it's worth it.

OpenCV's color converter is written as a OpenCL kernel and executes the conversion in parallel with many threads and super high efficiency, thus it's able to convert color data way faster than a Numpy-view-trick-and-Copy operation would be able to do. But this speed gain relies on `cv2.cvtColor()` being _given_ already-contiguous data and modifying it in-place without having to copy/allocate any new RAM. In that scenario, it's _much_ faster than pure Numpy `img = img[..., [2, 1, 0]]; img = img.copy()` tricks.

However, if we pass a non-contiguous data blob to `cv2.cvtColor()`, it internally calls Numpy's "copy as contiguous array" function (to make the image data usable/readable by OpenCV) before it converts the color. Basically, RAM allocation and data copying will happen _whenever_ you call OpenCV functions with _any_ non-contiguous input data. So it's the same total time cost as manually invoking a Numpy `img.copy()`-call first before you call OpenCV's `cv2.cvtColor()`.

Because the data from the Windows API is already messed up, there may not be any time saving for you, and Numpy is probably even faster... it depends on what your code is doing. Here are the two methods compared:

- Method 1: Having `frame` which is non-contiguous Numpy image data. Then doing Numpy `frame[..., [2, 1, 0]]` to change the channel order (takes like a thousandth of a millisecond), and then doing Numpy `.copy()` to copy the rearranged colors into a new, clean, contiguous Numpy object. The cost of this is basically the same as the cost of any other Numpy copy that uses complex strides.
- Method 2: Having `frame` which is non-contiguous Numpy image data. Giving it to `cv2.cvtColor()` which then internally calls Numpy `.copy()` to fix the data layout (using the same slow and complex stride handling inside Numpy as in Method 1), and then OpenCV modifies THAT brand new allocated RAM to change the color order very efficiently. But, in the end, this may be slower than Method 1, because it involves first doing a Numpy "complex array stride" `.copy()`, and THEN an OpenCV re-arrange of THAT new data.

In short, Method 1 is "Have complex strides and tell Numpy to do a clean copy". And Method 2 is "Have complex strides and tell Numpy to do a clean copy, and THEN modify that clean RAM to change the color order".

In the end, I think Method 1 is best for this library. Method 2 will just be extra work since the data is already messed up and non-contiguous. So just keep doing the numpy "view manipulation" stride magic, which is "instantaneous".

Then simply educate users that they need to call `.copy()` on the screenshot to force Numpy to generate clean, contiguous data if they care about performance, since all image processing libraries either internally do a copy if given non-contiguous data, *or* they interpret the non-contiguous data with complex, slow loops. So it's basically always faster to clean up the screenshot data's RAM layout via `.copy()` if you're going to be giving the data to _any_ libraries.

It's also good optimization-wise that the library *doesn't* internally forcibly call Numpy's `.copy()`. Because the user _may_ only want a small _portion_ of the screen, in which case it's faster if the user gets the messy Numpy image "view manipulated" array from D3DShot, and then the user makes yet another slice of that, such as a "100 x 100 square in the center of the screenshot", and THEN the user calls `copy()` on THAT tiny view (to create a contiguous buffer of JUST the data they care about), which would be way faster than forcibly copying the entire screen's image buffer every time.

**TL;DR:** D3DShot should be using Numpy "view" magic tricks to generate the final data "interpretation" instructions. And the *users* should then manually call `.copy()` on that frame to optimize the screenshot data when it's necessary for various image library call performance.
